File: packages/node/src/sinks/bigquery.ts
import type { Sink } from './types.js';
import type { AnyEvent } from '../events.js';

export interface BigQuerySinkOptions {
  /** An existing `@google-cloud/bigquery` client to reuse. If omitted, one is created from Application Default Credentials. */
  client?: unknown;
  /** BigQuery project id. Defaults to the client library's own ADC-based resolution. */
  projectId?: string;
  dataset?: string;
  toolCallTable?: string;
}

interface BigQueryTable {
  insert(rows: Record<string, unknown>[]): Promise<unknown>;
}
interface BigQueryClient {
  dataset(name: string): { table(name: string): BigQueryTable };
}

/**
 * Streams rows into the table defined by schema/events.md's BigQuery DDL.
 * Requires the optional peer dependency `@google-cloud/bigquery` —
 * dynamically imported so it isn't required unless this sink is actually
 * used. Credentials come from Application Default Credentials, same as the
 * client library's own defaults.
 */
export function bigquerySink(options: BigQuerySinkOptions = {}): Sink {
  const dataset = options.dataset ?? 'mcpsignals';
  const toolCallTable = options.toolCallTable ?? 'tool_call';

  let clientPromise: Promise<BigQueryClient> | undefined;

  async function getClient(): Promise<BigQueryClient> {
    if (!clientPromise) {
      clientPromise = (async () => {
        if (options.client) {
          return options.client as BigQueryClient;
        }
        const mod = await import('@google-cloud/bigquery');
        const BigQuery = mod.BigQuery;
        return new BigQuery(
          options.projectId ? { projectId: options.projectId } : undefined
        ) as unknown as BigQueryClient;
      })();
    }
    return clientPromise;
  }

  return {
    async write(events: AnyEvent[]): Promise<void> {
      const client = await getClient();
      const toolCalls = events.filter(
        (e): e is Extract<AnyEvent, { event_type: 'tool_call' }> => e.event_type === 'tool_call'
      );
      if (toolCalls.length > 0) {
        await client
          .dataset(dataset)
          .table(toolCallTable)
          .insert(
            toolCalls.map(e => ({
              ts: e.ts.toISOString(),
              server_name: e.server_name,
              server_version: e.server_version,
              tool_name: e.tool_name,
              session_id: e.session_id,
              agent_id: e.agent_id,
              client_name: e.client_name,
              client_version: e.client_version,
              user_id: e.user_id,
              org_id: e.org_id,
              duration_ms: e.duration_ms,
              success: e.success,
              error_kind: e.error_kind,
              error_message: e.error_message,
              request_bytes: e.request_bytes,
              response_bytes: e.response_bytes,
              arguments: e.arguments ? JSON.stringify(e.arguments) : null,
              intent: e.intent,
              transport: e.transport
            }))
          );
      }
    }
  };
}

File: packages/python/src/mcpsignals/sinks/bigquery.py
"""BigQuery sink. Requires the `bigquery` extra (google-cloud-bigquery).
Credentials come from Application Default Credentials, same as the client
library's own defaults - never an invented config file. The client library
is synchronous, so writes run in a thread via asyncio.to_thread and never
block the event loop.
"""

import asyncio
import dataclasses

from mcpsignals.events import SessionSummaryEvent, ToolCallEvent


class BigQuerySink:
    def __init__(self, dataset: str, project: str | None = None, client=None):
        self._dataset = dataset
        self._project = project
        self._client = client

    def _get_client(self):
        if self._client is None:
            from google.cloud import bigquery  # local import: don't require the SDK unless used

            self._client = bigquery.Client(project=self._project)
        return self._client

    def _row(self, event) -> dict:
        row = dataclasses.asdict(event)
        row.pop("event_type", None)
        row["ts"] = event.ts.isoformat() if event.ts else None
        return row

    def _insert_sync(self, table: str, rows: list[dict]) -> None:
        client = self._get_client()
        table_ref = f"{self._dataset}.{table}"
        errors = client.insert_rows_json(table_ref, rows)
        if errors:
            raise RuntimeError(f"bigquery insert errors for {table_ref}: {errors}")

    async def write(self, events: list[ToolCallEvent | SessionSummaryEvent]) -> None:
        tool_calls = [self._row(e) for e in events if isinstance(e, ToolCallEvent)]
        summaries = [self._row(e) for e in events if isinstance(e, SessionSummaryEvent)]

        if tool_calls:
            await asyncio.to_thread(self._insert_sync, "tool_call", tool_calls)
        if summaries:
            await asyncio.to_thread(self._insert_sync, "session_summary", summaries)
